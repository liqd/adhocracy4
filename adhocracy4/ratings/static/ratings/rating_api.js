import api from '../../../static/api'

/**
 * @param {number} number - the rating value, can be -1, 0 or 1
 * @param {string} objectId - the object to be rated
 * @param {string} contentTypeId - the content type id of the object

 * @returns {Promise<Array>} - an array with the rating data (number of negative
 *    and positive ratings) and the user rating data (value and id and userHasRating)
 *    and the complete response data from the API.
 */
export async function createRating (number, objectId, contentTypeId) {
  try {
    const data = await api.rating.add({
      urlReplaces: {
        objectPk: objectId,
        contentTypeId
      },
      value: number
    })

    return [{
      positive: data.meta_info.positive_ratings_on_same_object,
      negative: data.meta_info.negative_ratings_on_same_object
    }, {
      userRating: data.meta_info.user_rating_on_same_object_value,
      userHasRating: true,
      userRatingId: data.id
    }, data]
  } catch (error) {
    // if the server responds with a 400 and an array with a single integer
    // the user has voted in the meantime. This handles the case and instead
    // uses the returned ID to modify the rating.
    if (error.status === 400 &&
        error.responseJSON.length === 1 &&
        Number.isInteger(parseInt(error.responseJSON[0]))
    ) {
      const userRatingId = parseInt(error.responseJSON[0])
      const [ratings, userState, data] = (await modifyRating(number, userRatingId))[0]
      return [
        ratings,
        {
          ...userState,
          userHasRating: true,
          userRatingId
        },
        data
      ]
    }
  }
}

/**
 * @param {number} number - the rating value, can be -1, 0 or 1
 * @param {number} id - the id of the users rating
 * @param {string} objectId - the object to be rated
 * @param {string} contentTypeId - the content type id of the object

 * @returns {Promise<Array>} - an array with the rating data (number of negative
 * and positive ratings) and the user rating data (value)
 * and the complete response data from the API.
 */
export async function modifyRating (number, id, objectId, contentTypeId) {
  const data = await api.rating.change({
    urlReplaces: {
      objectPk: objectId,
      contentTypeId
    },
    value: number
  }, id)
  return [
    {
      positive: data.meta_info.positive_ratings_on_same_object,
      negative: data.meta_info.negative_ratings_on_same_object
    },
    {
      userRating: data.meta_info.user_rating_on_same_object_value,
      userHasRating: number !== 0
    },
    data
  ]
}

/**
 * Helper function to easily create OR modify a rating
 *
 * @param {number} number - the rating value, can be -1, 0 or 1
 * @param {string} objectId - the object to be rated
 * @param {string} contentTypeId - the content type id of the object
 * @param {number} [id] - the id of the users rating

 * @returns {Promise<Array>} - an array with the rating data (number of negative
 * and positive ratings) and the user rating data (value)
 * and the complete response data from the API.
 */
export async function createOrModifyRating (number, objectId, contentTypeId, id) {
  if (id) {
    return await modifyRating(number, id, objectId, contentTypeId)
  } else {
    return await createRating(number, objectId, contentTypeId)
  }
}
