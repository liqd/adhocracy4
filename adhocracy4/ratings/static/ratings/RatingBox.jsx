import React, { useCallback, useMemo, useState } from 'react'
import django from 'django'

import { createOrModifyRating } from './rating_api'
import RatingButton from './RatingButton'
import config from '../../../static/config'

const translations = {
  likes: django.gettext('Likes'),
  dislikes: django.gettext('Dislikes')
}

export const getRedirectUrl = (id) => config.getLoginUrl() + (id ? encodeURIComponent('?comment=' + id) : '')

const RatingBox = ({
  positiveRatings,
  negativeRatings,
  userHasRating,
  userRating,
  userRatingId,
  isReadOnly,
  contentType,
  objectId,
  authenticatedAs,
  isComment,
  render
}) => {
  const [ratings, setRatings] = useState({ negative: negativeRatings, positive: positiveRatings })
  const [userRatingData, setUserRatingData] = useState({ userHasRating, userRating, userRatingId })

  const clickHandler = useCallback(async (number) => {
    if (!authenticatedAs) {
      const redirectId = isComment ? objectId : null
      window.location.href = getRedirectUrl(redirectId)
    }

    if (isReadOnly) {
      return null
    }

    const [ratings, newUserRatingData] = await createOrModifyRating(number, objectId, contentType, userRatingData.userRatingId)

    setRatings(ratings)
    setUserRatingData({ ...userRatingData, ...newUserRatingData })
  }, [authenticatedAs, objectId, contentType, userRatingData, isComment, isReadOnly])

  const customChildren = useMemo(() => {
    if (render && typeof render === 'function') {
      return render({ ratings, userRatingData, isReadOnly, clickHandler })
    }
    return null
  }, [clickHandler, isReadOnly, ratings, render, userRatingData])

  // return either custom html from the render prop or the default
  return customChildren ?? (
    <div className="rating" data-testid="rating-box">
      <RatingButton
        isReadOnly={isReadOnly}
        onClick={clickHandler}
        active={userRatingData.userRating === 1}
        rating={1}
        authenticatedAs={authenticatedAs}
      >
        <i className="fa fa-thumbs-up" aria-hidden="true" />{' '}
        {ratings.positive}{' '}
        <span className="rating__label">{translations.likes}</span>
      </RatingButton>
      <RatingButton
        isReadOnly={isReadOnly}
        onClick={clickHandler}
        active={userRatingData.userRating === -1}
        rating={-1}
        authenticatedAs={authenticatedAs}
      >
        <i className="fa fa-thumbs-down" aria-hidden="true" />{' '}
        {ratings.negative}{' '}
        <span className="rating__label">{translations.dislikes}</span>
      </RatingButton>
    </div>
  )
}

export default RatingBox
