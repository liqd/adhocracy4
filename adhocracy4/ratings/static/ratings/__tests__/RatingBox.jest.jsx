import React from 'react'
import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import RatingBox from '../RatingBox'
import { createOrModifyRating } from '../rating_api'

jest.mock('../../../../static/config', () => {
  return {
    getLoginUrl: jest.fn().mockReturnValue('/login')
  }
})

jest.mock('../rating_api')

const getData = ({ pos = 1, neg = 0, id = 2, userRating = null } = {}) => {
  const mockData = {
    meta_info: {
      positive_ratings_on_same_object: pos,
      negative_ratings_on_same_object: neg,
      user_rating_on_same_object_value: userRating,
      user_rating_on_same_object_id: 2
    },
    id
  }

  return [{
    positive: mockData.meta_info.positive_ratings_on_same_object,
    negative: mockData.meta_info.negative_ratings_on_same_object
  }, {
    userRating,
    userHasRating: userRating !== null,
    userRatingId: userRating !== null ? 2 : null
  }, mockData]
}

const props = {
  positiveRatings: 1,
  negativeRatings: 0,
  userHasRating: false,
  userRating: null,
  userRatingId: null,
  authenticatedAs: 'someone',
  contentType: 14,
  objectId: 10
}

describe('RatingBox', () => {
  const originalWindow = { ...window }
  const originalLocation = window.location

  beforeAll(() => {
    delete window.location
    window.location = { href: '' }
  })

  beforeEach(() => {
    createOrModifyRating.mockClear()
    window.adhocracy4 = { getCurrentPath: jest.fn(() => '') }
    window.location.href = ''
  })

  afterAll(() => {
    window.location = originalLocation
    Object.assign(window, originalWindow)
  })

  test('component renders', () => {
    render(<RatingBox />)
    const ratings = screen.getByTestId('rating-box')
    const buttons = screen.getAllByRole('button')

    expect(ratings).toBeTruthy()
    expect(buttons).toHaveLength(2)
    expect(buttons[0]).toHaveAccessibleName(/like/i)
    expect(buttons[1]).toHaveAccessibleName(/dislike/i)
  })
  test('updates on click', async () => {
    render(<RatingBox {...props} />)
    const button = screen.getByRole('button', { name: /1 like/i })
    expect(screen.getByRole('button', { name: /0 dislike/i })).toBeInTheDocument()

    createOrModifyRating.mockReturnValue(getData({ pos: 2, userRating: 1 }))
    fireEvent.click(button)
    await waitFor(() => expect(createOrModifyRating).toHaveBeenCalledTimes(1))
    expect(createOrModifyRating).toHaveBeenCalledWith(1, 10, 14, null)
    expect(screen.getByRole('button', { name: /0 dislike/i })).toBeInTheDocument()
    expect(button).toHaveAccessibleName(/2 likes/i)
  })
  test('redirects when not logged in', async () => {
    render(<RatingBox {...props} authenticatedAs={null} />)

    const button = screen.getByRole('button', { name: /1 like/i })
    fireEvent.click(button)

    await waitFor(() => expect(window.location.href).toEqual('/login'))
  })

  test('redirects when not logged in and readonly', async () => {
    render(<RatingBox {...props} authenticatedAs={null} isReadOnly />)

    const button = screen.getByRole('button', { name: /1 like/i })
    fireEvent.click(button)

    await waitFor(() => expect(window.location.href).toEqual('/login'))
  })

  test('redirects to login and specific comment in next parameter', async () => {
    const testProps = {
      ...props,
      authenticatedAs: null,
      isReadOnly: true,
      isComment: true,
      objectId: 2
    }

    render(<RatingBox {...testProps} />)

    const button = screen.getByRole('button', { name: /1 like/i })
    expect(button).toBeInTheDocument()
    fireEvent.click(button)

    await waitFor(() => {
      expect(window.location.href).toEqual('/login%3Fcomment%3D2')
    })
  })
})

describe('RatingBox custom render', () => {
  beforeEach(() => {
    createOrModifyRating.mockClear()
  })
  test('component renders with custom children', async () => {
    render(
      <RatingBox
        {...props} render={
        ({ ratings, userRatingData, isReadOnly, clickHandler }) => (
          <>
            <div data-testid="custom-rating-box">
              {JSON.stringify({ ratings, userRatingData, isReadOnly })}
              <button onClick={() => {
                clickHandler(-1)
              }}
              >
                dislike
              </button>
            </div>
          </>
        )
      }
      />
    )

    const ratings = screen.getByTestId('custom-rating-box')
    expect(ratings).toMatchSnapshot()

    const button = screen.getByRole('button', { name: /dislike/ })
    createOrModifyRating.mockReturnValue(getData())
    fireEvent.click(button)
    await waitFor(() => expect(createOrModifyRating).toHaveBeenCalledTimes(1))
    expect(createOrModifyRating).toHaveBeenCalledWith(-1, 10, 14, null)
  })
})
