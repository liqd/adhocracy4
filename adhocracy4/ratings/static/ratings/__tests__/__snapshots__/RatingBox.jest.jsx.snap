// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingBox custom render component renders with custom children 1`] = `
<div
  data-testid="custom-rating-box"
>
  {"ratings":{"negative":0,"positive":1},"userRatingData":{"userHasRating":false,"userRating":null,"userRatingId":null}}
  <button>
    dislike
  </button>
</div>
`;
